app/dashboard/src/layouts/AssetsTable.tsx

```diff
@@ -412,42 +322,249 @@ export default function AssetsTable(props: AssetsTableProps) {
     ReadonlySet<backendModule.AssetId>
   > | null>(null)
-  const [, setQueuedAssetEvents] = React.useState<assetEvent.AssetEvent[]>([])
   const nameOfProjectToImmediatelyOpenRef = React.useRef(initialProjectName)
   const rootDirectoryId = React.useMemo(
     () => backend.rootDirectoryId(user) ?? backendModule.DirectoryId(''),
     [backend, user],
   )
-  const [assetTree, setAssetTree] = React.useState<assetTreeNode.AnyAssetTreeNode>(() => {
-    const rootParentDirectoryId = backendModule.DirectoryId('')
-    return AssetTreeNode.fromAsset(
-      backendModule.createRootDirectoryAsset(rootDirectoryId),
+
+  const rootParentDirectoryId = backendModule.DirectoryId('')
+  const rootDirectory = React.useMemo(
+    () => backendModule.createRootDirectoryAsset(rootDirectoryId),
+    [rootDirectoryId],
+  )
+
+  const enableAssetsTableBackgroundRefresh = useFeatureFlag('enableAssetsTableBackgroundRefresh')
+  const assetsTableBackgroundRefreshInterval = useFeatureFlag(
+    'assetsTableBackgroundRefreshInterval',
+  )
+  /**
+   * The expanded directories in the asset tree.
+   */
+  const [expandedDirectoryIds, setExpandedDirectoryIds] = React.useState<
+    backendModule.DirectoryId[]
+  >(() => [rootDirectory.id])
+
+  const expandedDirectoryIdsSet = React.useMemo(
+    () => new Set(expandedDirectoryIds),
+    [expandedDirectoryIds],
+  )
+
+  const createProjectMutation = useMutation(
+    backendMutationOptions(backend, 'createProject', {
+      meta: { invalidates: [['listDirectory', backend.type]] },
+    }),
+  )
+  const duplicateProjectMutation = useMutation(
+    backendMutationOptions(backend, 'duplicateProject', {
+      meta: { invalidates: [['listDirectory', backend.type]] },
+    }),
+  )
+  const createDirectoryMutation = useMutation(
+    backendMutationOptions(backend, 'createDirectory', {
+      meta: { invalidates: [['listDirectory', backend.type]] },
+    }),
+  )
+  const createSecretMutation = useMutation(
+    backendMutationOptions(backend, 'createSecret', {
+      meta: { invalidates: [['listDirectory', backend.type]] },
+    }),
+  )
+  const updateSecretMutation = useMutation(
+    backendMutationOptions(backend, 'updateSecret', {
+      meta: { invalidates: [['listDirectory', backend.type]] },
+    }),
+  )
+  const createDatalinkMutation = useMutation(
+    backendMutationOptions(backend, 'createDatalink', {
+      meta: { invalidates: [['listDirectory', backend.type]] },
+    }),
+  )
+  const uploadFileMutation = useMutation(
+    backendMutationOptions(backend, 'uploadFile', {
+      meta: { invalidates: [['assetVersions'], ['listDirectory', backend.type]] },
+    }),
+  )
+  const getProjectDetailsMutation = useMutation(
+    backendMutationOptions(backend, 'getProjectDetails'),
+  )
+
+  const directories = useQueries({
+    // we fetch only expanded directories
+    // this is an optimization technique to reduce the amount of data to fetch
```

Comment:
- should be proper sentences
- not sure whether i'd call it an "optimzation technique", exactly... if we fetch *all* directories then we risk having to fetch a potentially unlimited number of directories (depending on e.g. org size)